Feature/layering

fmontei opened the pull request:
This PR adds the following functionality to Deckhand:
- Add layering logic to Deckhand's engine, allowing documents to be layered together, as per
the requirements in the design document.
- Add support for merge, delete and replace actions.

New unit tests were also added:
- Add unit tests to test basic merge actions.
- Add unit tests to handle more complicate layering scenarios.
- Add unit tests to handle basic scenarios included in design document.

Commit:
Add base deckhand exception.

Commit:
Refactor some code.

Commit:
Add more docstrings/comments.

Commit:
Update the document's data if it is concrete.

Commit:
Add UnsupportedActionMethod exception.

Commit:
Resolve multiple bugs/wrong assumptions

This commit makes the following changes:
  - change merge to be a "deep merge" meaning that dictionary
    entries are recursively merged together; merge conflicts are
    resolved by prioritizing the current document's value over
    the parent document's value
  - replace without the entry in the current document will copy
    over entry (if exists) from parent
  - merge wi thout the entry in current document will copy
    over entry (if exists) from parent
  - reworte many unit tests and added some comments to make them
    more understandable

Commit:
Fixed spacing/some comments.

Commit:
Fix previously uncommented unit test.

Commit:
Implemented initial logic for layering.

mark-burnett commented on deckhand/tests/unit/engine/test_document_layering.py:
```diff
@@ -0,0 +1,392 @@
+            self._test_layering(expected[idx], documents)
+
+    def test_layering_method_merge(self):
+        expected = [
```
I think my expectations diverge a bit from this test.  @bryan-strassner and I have discussed having an implicit action of `{path: ., method: merge}` at the beginning of every layering definition, so that by default we include all the data from the child document, but essentially I would only expect to see the child data that's explicitly called out to be included in the render.

Case `path: .` is as I expect.

Case `path: .a` I would expect  `{a: {x: 7, y: 2, z: 3}, c: 9}`

Case `path: .b` I would expect `{a: {x: 1, y: 2}, b: 4, c: 9}`

Case `path: .c` I would expect `{a: {x: 1, y: 2}, c: 9}` -- Actually, maybe I would expect a validation error, since there is no `c` key in the child.

If we don't add the implicit merge as the first action, then I think we should consider a validation that looks for data in the child document that would not be included.  I think it's not very intuitive to have data in the child document that doesn't make it in, so maybe we should add the implicit action.  Thoughts?

fmontei replied:
@mark-burnett  I rewrote my unit tests to make it easier to keep track of changes, so for:

```python

            kwargs = {
                "_GLOBAL_DATA_": {"data": {"a": {"x": 1, "y": 2}, "c": 9}},
                "_SITE_DATA_": {"data": {"a": {"x": 7, "z": 3}, "b": 4}},
                "_SITE_ACTIONS_": {
                    "actions": [{"method": "merge", "path": ".a"}]}
            }
```

Does this translate to:

1) Keep all the site data but merge the site data in path ".a" with the global data in path ".a". Or:
2) Only keep the result of merging the site data in path ".a" with the global data in path ".a".

The second is not self-evident because "merge" implies "this plus this" versus "this plus this...except this".

Want to clarify this before making changes everywhere.

fmontei replied:
Also, why is "Case `path: .a` I would expect `{a: {x: 7, y: 2, z: 3}, c: 9}`" true? Why is "c" included in the result? Why not "b"? If the merge path is ".a" why is "b" being dropped from the site data and "c" being included from the global data?

I'd imagine that should only be possible if:

 ```python

    kwargs = {
                "_GLOBAL_DATA_": {"data": {"a": {"x": 1, "y": 2}, "c": 9}},
                "_SITE_DATA_": {"data": {"a": {"x": 7, "z": 3}, "b": 4}},
                "_SITE_ACTIONS_": {
                    "actions": [{"method": "merge", "path": ".a"}, {"method": "merge", "path": ".c"}, {"method": "delete", "path": ".b"}]}
            }
```

But as you say above, this would be impossible, since if ".c" is not in the child, you expect a validation error. So how do you get `"c": 9` into the child other than doing a "merge" with a path of `"."`?

Commit:
Refactor layering unit tests.

Commit:
Finish refactoring unit tests.

Commit:
TEMP - Updating all concrete layers.

Commit:
Update all concrete documents in a chain. Add unit tests for such cases.

Commit:
Remove misleading comment.

Commit:
Reverse layering order.

Commit:
More tests.

Commit:
Update some comments.

Commit:
Remove dead code.

Commit:
More negative/positive tests for different scenarios.

Commit:
temp

Commit:
Fix a few bugs, refactor some code

This commit makes layering work for more complex layer hierarchies,
like:
  - 1 global -> 2 sites
  - 1 global -> 2 regions -> 1 site each
  - 2 globals -> 1 site each (this case should probably be disallowed)

Added some unit tests to tests these cases. Before the unit tests
were not "randomized" enough to notice bugs. Fixed that with this
commit. Also updated Document wrapper object to provide a higher
level of abstraction around certain functionality, like deleting
an item from a dictionary or retrieving all nested children for
a document.

Commit:
Another test.

Commit:
Fix reference bug and fix concrete/abstract bug.

Commit:
Line length.

Commit:
Fix some comments after resolving merge conflicts.

Commit:
Remove deprecated unit test.

Commit:
Another negative test.

Commit:
Add factory for generating document templates for testing.

alanmeadows reviewed (approved):
Felipe, this is an impressive set of Python work.

fmontei commented on deckhand/engine/document.py:
```diff
@@ -0,0 +1,103 @@
+        self._inner = data
+
+    @property
+    def all_data(self):
```
Better to call this ``to_dict()``.

fmontei commented on deckhand/engine/layering.py:
```diff
@@ -0,0 +1,292 @@
+            policy document).
+        """
+        # ``rendered_data_by_layer`` agglomerates the set of changes across all
+        # actions across each layer for a specific document. Use copy
```
The "Use copy..." comment is no longer valid and should be removed.

fmontei replied:
Actually, it should be moved to L125 below.

fmontei commented on deckhand/engine/layering.py:
```diff
@@ -0,0 +1,292 @@
+            if 'children' in doc:
+                del doc['children']
+
+        return [d.all_data for d in self.layered_docs]
```
This'll be a little easier to understand if it's ``d.to_dict() for d in self.layered_docs``.

fmontei commented on deckhand/engine/layering.py:
```diff
@@ -0,0 +1,292 @@
+              ``overall_data`` (in both the parent and child documents).
+        """
+        method = action['method']
+        if method not in self.SUPPORTED_METHODS:
```
Add TODO saying this should defer to schema validation. Validation should've been done by now,

fmontei commented on deckhand/engine/layering.py:
```diff
@@ -0,0 +1,292 @@
+            the topmost layer defined by the `layerOrder`.
+
+        :returns: Ordered list of documents that need to be layered. Each
+            document contains a "parent" in addition to original data. The
```
This comment is deprecated. "Parent" is no longer added -- rather "children" is.

fmontei commented on deckhand/engine/layering.py:
```diff
@@ -0,0 +1,292 @@
+            filter(lambda x: 'layeringDefinition' in x['metadata'],
+                self.documents))
+
+        all_children = collections.Counter()
```
Add comment: "all_children is counter utility for verifying that each document has exactly one parent."

Commit:
Fix up some comments, update deprecated comments.

fmontei commented:
@alanmeadows Thanks. Since you approved, just updated some comments before merge. Just need @mark-burnett to look at this to make sure the logic is correct when he's back from vacation.

Commit:
Test for both abstract.

Commit:
Resolved merge conflicts.

Commit:
Prevent documents with different schemas from being layered.

Commit:
Fix test case.

fmontei commented:
Moved to Gerrithub: https://review.gerrithub.io/#/c/373140/
